daphne_worker/src/durable/report_store.rs

```diff
@@ -111,8 +113,50 @@ impl DurableObject for ReportStore {
                     .size()
                     == 0;
 
-                // Check if this bucket is now empty, and if so, remove it from the agg job queue.
-                if empty {
+                // BUG(issue#73) There is a race condition between DURABLE_REPORT_STORE_GET_PENDING
+                // and DURABLE_REPORT_STORE_PUT_PENDING that could prevent a report from getting
+                // processed.
+                //
+                //   1. DURABLE_REPORT_STORE_PUT_PENDING writes a report to storage and checks if
+                //      "agg_job" is set. If not, it sets "agg_job" and enqueues the job.
+                //
+                //   2. DURABLE_REPORT_STORE_GET_PENDING drains a number of reports from storage,
+                //      then checks if storage is empty. If so, it deletes "agg_job" if it exists
+                //      and dequeues the job.
+                //
+                // Consider the following sequence of storage transactions (suppose "agg_job" is
```

Comment:
Hmm. I'm no longer certain of the order of events I suggest above, due to the following from [the CF workers blog post](https://blog.cloudflare.com/durable-objects-easy-fast-correct-choose-three):

> Input gates: While a storage operation is executing, no events shall be delivered to the object except for storage completion events. Any other events will be deferred until such a time as the object is no longer executing JavaScript code and is no longer waiting for any storage operations. We say that these events are waiting for the "input gate" to open.

(N.B. I can't find the input/output gate rules in any official documentation, but I assume being part of an official CF blog post is good enough.)

IIUC these "input gates" should stop a concurrent `DURABLE_REPORT_STORE_GET_PENDING` & `DURABLE_REPORT_STORE_PUT_PENDING` from being possible. I suppose if there are eventual-consistency issues to be concerned with, a similar race could still happen, but I _think_ DOs don't have this by virtue of running all storage operations in a single instance.

This text from the same blog post is worrying, but I'm not sure that we're triggering this behavior (or how to be sure if we are or not):

> Two calls to getUniqueNumber() are initiated by the same event. The application does not await the first call before starting the second, so the two calls end up running concurrently. Our special rule doesn't protect us here, because there is no incoming event that can be deferred between when the two calls are made. From the system's point of view, there's no way to distinguish this code from code which legitimately decided to perform two storage operations in parallel.
> 
> As such, in this case, the two calls to getUniqueNumber() will interfere with each other.

I'm going to stare at this for awhile longer, I am now extremely curious what might be going on.

Reply:
The [original Durable Objects announcement post](https://blog.cloudflare.com/introducing-workers-durable-objects) confirms that DO's are immediately consistent (ie **not** eventually consistent):

> [Durable Objects] are extremely well-suited to workloads requiring transactional guarantees and immediate consistency.

Reply:
The [miniflare 2.0 announcement](https://noise.getoto.net/2022/01/07/miniflare-2-0-fully-local-development-and-testing-for-workers) listed implementation of Durable Object input/output gates as a feature, so miniflare supports this behavior (modulo bugs).

Reply:
Thanks for digging @BranLwyd! After conferring with SMEs internally, in all likelihood miniflare is actually matching prod here. What I've learned is that the call to LeaderAggregationJobQueue from ReportStore breaks the guarantee that storage transactions won't get interleaved across transactions. I'll have to figure out a different design.

In the meantime, I will land this once I can get review from my side on Monday, in order to unblock you. I'll work on a permanent fix next week.

Reply:
I confirmed with some logging that, during a failure test-case, the same `ReportStore` object instance entered both the `PUT_PENDING` & `GET_PENDING` operations concurrently. This is consistent with the order of events suggested in the opening comment of this chain, but apparently inconsistent with how input gates work.

But thinking about this some more, I think the input gate protections don't matter because they only apply "while a storage operation is executing." Current code always awaits storage operations immediately. I think this means that when the `GET_PENDING` operation sends a fetch request to the `LeaderAggregationJobQueue` DO, other operations against the `ReportStore` DO are allowed to run concurrently.

Logging off for now but I'm going to let this percolate. We may just need to wrap these operations in an explicit transaction instead of relying on DO's input gate magic, but I haven't yet worked out if that sufficiently protects the paired, non-transactional operations on the `ReportStore` & `LeaderAggregationJobQueue` DOs when a `ReportStore` becomes empty/non-empty.

Reply:
This is one possible solution, but unfortunately is not implemented by the WASM wrapper code:  `blockConcurrencyWhile` https://developers.cloudflare.com/workers/runtime-apis/durable-objects/

However I think there's a different (and maybe better) design here: Have PUT_PENDING check if there a sufficient number (10, say) reports in the bucket to queue an agg job. If so, remove them, then send a request to LeaderAggregateJobQueue that contains these reports. Then we can get rid of GET_PENDING altogether.